### PR TITLE
[Issue #359]  Direct Surrogate Support

### PR DESCRIPTION
**Pull request by sujith71955**
The below file changes is basically done for supporting direct query filter support, since the
filter value resolving mechanism is increasing the code has been re factored also for accommodating such extensions

### CODE REVIEW

**Comment by sujith71955** [comment]
retest this please

File: core/src/main/java/org/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java [hunk 1]
```diff
@@ -118,12 +123,14 @@
           + e.getMessage());
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java]: Error Message logging is not required

File: core/src/main/java/org/carbondata/query/carbon/executor/util/QueryUtil.java [hunk 2]
```diff
@@ -278,19 +278,24 @@ public static void extractDimensionsAndMeasuresFromExpression(
...
-      if (queryDimensions.get(i).getDimension().getEncoder().contains(Encoding.DICTIONARY)) {
+      List<Encoding> encodingList = queryDimensions.get(i).getDimension().getEncoder();
+      if (CarbonUtil.hasEncoding(encodingList, Encoding.DICTIONARY) && !CarbonUtil
+          .hasEncoding(encodingList, Encoding.DIRECT_DICTIONARY)) {
         dictionaryDimensionFromQuery.add(queryDimensions.get(i).getDimension().getColumnId());
       }
     }
     for (int i = 0; i < dimAggInfo.size(); i++) {
-      if (dimAggInfo.get(i).getDim().getEncoder().contains(Encoding.DICTIONARY)) {
+      List<Encoding> encodingList = dimAggInfo.get(i).getDim().getEncoder();
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/carbon/executor/util/QueryUtil.java]: write a comment, skipping is done for only dictionary lookup

File: core/src/main/java/org/carbondata/query/expression/ExpressionResult.java [hunk 3]
```diff
@@ -112,7 +112,7 @@ public String getString() {
         if (value instanceof Timestamp) {
           return parser.format((Timestamp) value);
         } else {
-          return parser.format(new Timestamp((long) value));
+          return parser.format(new Timestamp((long) value/1000));
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/expression/ExpressionResult.java]: % 1000 case in getInt function should throw execption

File: integration/spark/src/test/scala/org/carbondata/spark/testsuite/directdictionary/TimestampDataTypeDirectDictionaryTest.scala [hunk 4]
```diff
@@ -19,19 +19,21 @@
 
 package org.carbondata.spark.testsuite.directdictionary
 
+import java.io.File
```
Comment (gvramana) [inline integration/spark/src/test/scala/org/carbondata/spark/testsuite/directdictionary/TimestampDataTypeDirectDictionaryTest.scala]: add _FT for ft testcases( confirm the syntax)

File: core/src/main/java/org/carbondata/core/keygenerator/directdictionary/DirectDictionaryKeyGeneratorFactory.java [hunk 5]
```diff
@@ -42,6 +45,26 @@ private DirectDictionaryKeyGeneratorFactory() {
   public static DirectDictionaryGenerator getDirectDictionaryGenerator(DataType dataType) {
     DirectDictionaryGenerator directDictionaryGenerator = null;
     switch (dataType) {
+      case TimestampType:
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/core/keygenerator/directdictionary/DirectDictionaryKeyGeneratorFactory.java]: This function is not required

File: core/src/main/java/org/carbondata/query/filter/resolver/resolverinfo/visitor/CustomTypeDictionaryVisitor.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      LOGGER.audit(e.getMessage());
+    }
+
+  }
+
+  private DimColumnFilterInfo getDirectDictionaryValKeyMemberForFilter(
+      AbsoluteTableIdentifier tableIdentifier, ColumnExpression columnExpression,
+      List<String> evaluateResultListFinal, boolean isIncludeFilter) {
+    List<Integer> surrogates = new ArrayList<Integer>(20);
+    DirectDictionaryGenerator directDictionaryGenerator = DirectDictionaryKeyGeneratorFactory
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/filter/resolver/resolverinfo/visitor/CustomTypeDictionaryVisitor.java]: If generator returns -1, how the condition to be treated.

File: core/src/main/java/org/carbondata/query/filter/resolver/resolverinfo/visitor/CustomTypeDictionaryVisitor.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+  /**
+   * This Visitor method is been used to resolve or populate the filter details
+   * by using custom type dictionary value, the filter membrers will be resolved using
+   * custom type function which will generate dictionary for the direct column type filter members
+   *
+   * @param visitableObj
+   * @param metadata
+   */
+  public void populateFilterResolvedInfo(DimColumnResolvedFilterInfo visitableObj,
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/filter/resolver/resolverinfo/visitor/CustomTypeDictionaryVisitor.java]: check if any abstract is required?

File: core/src/main/java/org/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java (hunk 1 above)
Comment (mohammadshahidkhan) [inline core/src/main/java/org/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java]: Removed the log message

File: core/src/main/java/org/carbondata/query/carbon/executor/util/QueryUtil.java (hunk 2 above)
Comment (mohammadshahidkhan) [inline core/src/main/java/org/carbondata/query/carbon/executor/util/QueryUtil.java]: fixed

File: integration/spark/src/test/scala/org/carbondata/spark/testsuite/directdictionary/TimestampDataTypeDirectDictionaryTest.scala (hunk 4 above)
Comment (mohammadshahidkhan) [inline integration/spark/src/test/scala/org/carbondata/spark/testsuite/directdictionary/TimestampDataTypeDirectDictionaryTest.scala]: fixed added _FT for functional and _UT for unit test

**Comment by sujith71955** [comment]
retest this please

**Comment by sujith71955** [comment]
retest this please

**Comment by sujith71955** [comment]
retest this please

**Pull request merged**